plugins/default/src/Ide/Plugin/Pragmas.hs

```diff
@@ -40,35 +43,38 @@ descriptor plId = (defaultPluginDescriptor plId)
 -- Pragma is added to the first line of the Uri.
 -- It is assumed that the pragma name is a valid pragma,
 -- thus, not validated.
-mkPragmaEdit :: Uri -> T.Text -> WorkspaceEdit
-mkPragmaEdit uri pragmaName = res where
-    pos = J.Position 0 0
-    textEdits = J.List
-      [J.TextEdit (J.Range pos pos)
-                  ("{-# LANGUAGE " <> pragmaName <> " #-}\n")
-      ]
-    res = J.WorkspaceEdit
-      (Just $ H.singleton uri textEdits)
-      Nothing
+mkPragmaEdit :: Uri -> T.Text -> Range -> WorkspaceEdit
+mkPragmaEdit uri pragmaName rng = res where
+    textEdits = J.List [J.TextEdit rng ("{-# LANGUAGE " <> pragmaName <> " #-}\n")]
+    res = J.WorkspaceEdit (Just . H.singleton uri $ textEdits) Nothing
 
 -- ---------------------------------------------------------------------
 -- | Offer to add a missing Language Pragma to the top of a file.
 -- Pragmas are defined by a curated list of known pragmas, see 'possiblePragmas'.
 codeActionProvider :: CodeActionProvider IdeState
-codeActionProvider _ state _plId docId _ (J.CodeActionContext (J.List diags) _monly) = do
+codeActionProvider lsp state _plId docId _ (J.CodeActionContext (J.List diags) _monly) = do
     let mFile = docId ^. J.uri & uriToFilePath <&> toNormalizedFilePath'
     pm <- fmap join $ runAction "addPragma" state $ getParsedModule `traverse` mFile
     let dflags = ms_hspp_opts . pm_mod_summary <$> pm
         -- Get all potential Pragmas for all diagnostics.
         pragmas = nubOrd $ concatMap (\d -> genPragma dflags (d ^. J.message)) diags
-    cmds <- mapM mkCodeAction pragmas
+        somh = startOfModuleHeader pm
+    loc <- if somh > 0
+            then do
+              contents <- LSP.getVirtualFileFunc lsp . toNormalizedUri $ docId ^. J.uri
+              text <- pure $ fmap VFS.virtualFileText contents
+              line <- pure . fromMaybe 0 $ findIndex (not . T.null) . take somh . T.lines =<< text
```

Comment:
The intention is when we have
```Haskell
#! /usr/bin/env nix-shell
#! nix-shell --pure -i runghc -p "haskellPackages.ghcWithPackages (hp: with hp; [ turtle ])"

{-# LANGUAGE ScopedTypeVariables #-}
module AfterShebang where

foo :: forall a. a -> a
foo = id @a
```

we get
```Haskell
#! /usr/bin/env nix-shell
#! nix-shell --pure -i runghc -p "haskellPackages.ghcWithPackages (hp: with hp; [ turtle ])"

{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeApplications #-}
module AfterShebang where

foo :: forall a. a -> a
foo = id @a
```
right? Then I have one question: Why do we need to check `contents`? I notice that `somh` itself is good enough...
Do you have any other cases that are not covered by `somh`?

Reply:
(BTW, this is why the tests fail)

Reply:
I could not figure out the case for contents as well. I mostly copied from `CodeAction.hs -> endOfModuleHeader`

Reply:
I used only `somh` now but still gives the same result locally due to the Nothing `pm`. Not sure about the actual cause of this.

Reply:
I used the `somh` only but can't figure out why it still adds pragma to the first line. Sorry about that

Reply:
Does it work in your local machine? GetParsedModuleWithComments errors probably come frome other tests, so could you check addPragma tests?

Reply:
I checked the Shebang test specifically with testy pattern. Does not work :(

Reply:
Give the same error as CI? Then probably one of the maybe is `Nothing`...

Reply:
As far as I could figure out `pm` is `Nothing` . Not sure what else I could do... Please point me to some other issue if this has no foreseeable solutions.

Reply:
That's weird, how did you check that?

Reply:
I am not that good with debugging Haskell. Had to resort to extreme measures :D

`mkPragmaEdit :: Uri -> T.Text -> Range -> Maybe ParsedModule -> WorkspaceEdit
mkPragmaEdit uri pragmaName rng pm = res where
    textEdits = J.List [J.TextEdit rng ("{-# LANGUAGE " <> pragmaName <> " #-}\n" <> T.pack (show pm) <> "\n")]
    res = J.WorkspaceEdit (Just . H.singleton uri $ textEdits) Nothing`

Reply:
Running the test with this gives a `Nothing` in place of `pm`

Reply:
OK... that sounds strange. When I tried with my local fix of this PR it gives valid result. Let me try that fix again.

Reply:
Thank you very much for being so patient

Reply:
@Ailrun please let me know if I can help in anything

Reply:
Sorry, I'm really busy for this week or two... maybe @jneira can also help? I'm not sure.

Reply:
will try

Reply:
Thanks @jneira ... means a lot